pkg/utils/cluster.go

```diff
@@ -15,39 +15,119 @@ import (
-	infraResourceName = "cluster"
+	infraResourceName  = "cluster"
+	workerNodeLabelKey = "node-role.kubernetes.io/worker"
+	masterNodeLabelKey = "node-role.kubernetes.io/master"
 )
 
+func getNodeRole(node corev1.Node) string {
+	for k := range node.Labels {
+		if k == workerNodeLabelKey {
+			return "worker"
+		} else if k == masterNodeLabelKey {
+			return "master"
+		}
+	}
+	return ""
+}
+
 func IsSingleNodeCluster(c client.Client) (bool, error) {
 	if os.Getenv("CLUSTER_TYPE") == ClusterTypeOpenshift {
-		return openshiftSingleNodeClusterStatus(c)
+		topo, err := openshiftControlPlaneTopologyStatus(c)
+		if err != nil {
+			return false, err
+		}
+		switch topo {
+		case configv1.SingleReplicaTopologyMode:
+			return true, nil
+		case configv1.HighlyAvailableTopologyMode:
+			return false, nil
+		}
+		return false, nil
 	}
 	return k8sSingleNodeClusterStatus(c)
 }
 
+// IsExternalControlPlaneCluster detects control plane location of the cluster.
+// On OpenShift, the control plane topology is configured in configv1.Infrastucture struct.
+// On kubernetes, it is determined by which node the sriov operator is scheduled on. If operator
+// pod is schedule on worker node, it is considered as external control plane.
+func IsExternalControlPlaneCluster(c client.Client) (bool, error) {
+	if os.Getenv("CLUSTER_TYPE") == ClusterTypeOpenshift {
+		topo, err := openshiftControlPlaneTopologyStatus(c)
+		if err != nil {
+			return false, err
+		}
+		switch topo {
+		case "External":
+			return true, nil
+		case configv1.HighlyAvailableTopologyMode:
+			return false, nil
+		case configv1.SingleReplicaTopologyMode:
+			return false, nil
+		}
+	} else {
+		role, err := operatorNodeRole(c)
+		if err != nil {
+			return false, err
+		}
+		switch role {
+		case "master":
+			return false, nil
+		case "worker":
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
 func k8sSingleNodeClusterStatus(c client.Client) (bool, error) {
 	nodeList := &corev1.NodeList{}
 	err := c.List(context.TODO(), nodeList)
 	if err != nil {
-		glog.Errorf("IsSingleNodeCluster(): Failed to list nodes: %v", err)
+		glog.Errorf("k8sSingleNodeClusterStatus(): Failed to list nodes: %v", err)
 		return false, err
 	}
 
 	if len(nodeList.Items) == 1 {
-		glog.Infof("IsSingleNodeCluster(): one node found in the cluster")
+		glog.Infof("k8sSingleNodeClusterStatus(): one node found in the cluster")
 		return true, nil
 	}
 	return false, nil
 }
 
-func openshiftSingleNodeClusterStatus(c client.Client) (bool, error) {
+// operatorNodeRole returns role of the node where operator is scheduled on
+func operatorNodeRole(c client.Client) (string, error) {
+	pod := &corev1.Pod{}
+	err := c.Get(context.TODO(), types.NamespacedName{Namespace: os.Getenv("NAMESPACE"), Name: os.Getenv("K8S_POD_NAME")}, pod)
+	if err != nil {
+		glog.Errorf("k8sIsExternalTopologyMode(): Failed to get operator pod: %v", err)
+		return "", err
+	}
+
+	nodeList := &corev1.NodeList{}
+	err = c.List(context.TODO(), nodeList)
```

Comment:
you gen get node by name no ? no need to list and iterate here

Reply:
fixed.